File: 0496 next-greater-element-i/solution.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        int st[10001];
        int top = -1;
        int map[10001];
        memset(map, -1, sizeof(map));
        for(int i:nums2) {
            while(top!=-1 && st[top] < i) {
                map[st[top--]] = i;
            }
            st[++top] = i;
        }
        vector<int> ans;
        for(int i:nums1) {
            ans.push_back(map[i]);
        }
        return ans;
    }
};

File: 0496 next-greater-element-i/solution.rs
use std::collections::HashMap;
impl Solution {
    pub fn next_greater_element(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        let mut st = Vec::new();
        let mut map = HashMap::new();
        for i in nums2 {
           while !st.is_empty() && *st.last().unwrap() < i {
                map.insert(st.pop().unwrap(), i);
            }
            st.push(i);
        }
       let mut ans = Vec::new();
       for i in nums1 {
            ans.push(*map.get(&i).unwrap_or(&-1));
        }
        ans
    }
}
